lib/awful/key.lua

```diff
@@ -99,6 +129,32 @@ local key = { mt = {}, hotkeys = {} }
 
 local reverse_map = setmetatable({}, {__mode="k"})
 
+--- The keygroups names.
+--
+-- It can be used instead of keygroup names.
+--
+-- Values associated to each property of this table are string:
+--
+-- - **NUMROW** = `"numrow"`: The row above the letters in the US PC-105/PC-104 keyboards and
+-- its derivative. This is usually the number 1-9 followed by 0.
+--
+-- - **ARROWS** = `"arrows"`: The Left/Right/Top/Bottom keys usually located right of the
+-- spacebar.
+--
+-- - **FKEYS** = `"fkeys"`: The keys F1 through F12 located at the topmost row of any
+-- keyboard, plus F13 through F35 on specialist keyboards.
+--
+-- - **NUMPAD** = `"numpad"`: The number keys on the keypad to the right of the letters and
+-- the arrow keys. Not present in every keyboard.
+--
+-- @table keygroup
+key.keygroup = {
+    NUMROW = 'numrow', -- The number row.
+    ARROWS = 'arrows', -- The directionnal arrows.
+    FKEYS = 'fkeys', -- The function keys.
+    NUMPAD = 'numpad', -- The numpad keys.
+}
```

Comment:
i wonder how this was working before

Reply:
You had to specify the literal string by yourself `"numrow" ` for example instead of `awful.key.keygroup.NUMROW`.

We still can use literal string, this is just a table that hold these strings. It makes the code feels less "random" as the user can now use this "enum".

It's exactly what was done before for the mouse buttons, we have the `awful.button.name` table to replace the use of mouse button indexes.

Reply:
in that case, i don't see in the diff the changes which are removing the old ones

Reply:
Hm... Sorry, it wasn't clear.

Currently, the doc have: https://awesomewm.org/apidoc/input_handling/awful.key.html#keygroup

```
keygroup N/A

    A group of keys.

    The valid keygroups are:

        numrow: The row above the letters in the US PC-105/PC-104 keyboards and its derivative. This is usually the number 1-9 followed by 0.
        arrows: The Left/Right/Top/Bottom keys usually located right of the spacebar.
        fkeys: The keys F1 through F12 located at the topmost row of any keyboard, plus F13 through F35 on specialist keyboards.
        numpad: The number keys on the keypad to the right of the letters and the arrow keys. Not present in every keyboard.


```

There is, however, no implementation in code for this property (https://github.com/awesomeWM/awesome/blob/master/lib/awful/key.lua#L27-L40). (BTW, it is documented as an object property (ie instance field) which is a lie as an instance doesn't have the field.) This property documentation is here to list the possible (string) values a user can use to refer a keygroup:

```lua
awful.key {
      modifiers = { "Mod4" },
      keygroup = "numrow", -- See here the literal string to refer to the numrow keygroup
      description = "only view tag",
      group = "tag",
      on_press = function (index) ... end
}
```

(The technical explanation to use a string here is that the table defining the keygroups uses these string as key (https://github.com/awesomeWM/awesome/blob/master/lib/awful/key.lua#L305-L363). So the string can be used latter as a key/an index to get the keygroup (https://github.com/awesomeWM/awesome/blob/master/lib/awful/key.lua#L377-L378))

The table I have defined here doesn't change any technical aspect of the current implementation. It only makes the doc less misleading and gives the user this "enum like table" I like:

```lua
awful.key {
      modifiers = { "Mod4" },
      keygroup = awful.key.keygroup.NUMROW, --  It feels less clumsy
      description = "only view tag",
      group = "tag",
      on_press = function (index) ... end
}
```

Note that you can still use the literal strings as parameter as the table basically only hold them.

Reply:
ah, i see, thanks a lot for the detailed clarification :)